src/Libraries/Text/Lexer/Tokenizer.idr

```diff
@@ -0,0 +1,148 @@
+|||                     form of `(startLine, startCol), (endLine, endCol)`.
+public export
+data StopReason = EndInput | NoRuleApply | ComposeNotClosing (Int, Int) (Int, Int)
+
+export
+Show StopReason where
+  show EndInput = "EndInput"
+  show NoRuleApply = "NoRuleApply"
+  show (ComposeNotClosing start end) = "ComposeNotClosing " ++ show start ++ " " ++ show end
+
+export
+Pretty StopReason where
+  pretty EndInput = pretty "EndInput"
+  pretty NoRuleApply = pretty "NoRuleApply"
+  pretty (ComposeNotClosing start end) = "ComposeNotClosing" <++> pretty start <++> pretty end
+
+tokenise : Lexer ->
+           Tokenizer a ->
+           (line, col : Int) -> List (WithBounds a) ->
+           List Char ->
+           (List (WithBounds a), (StopReason, Int, Int, List Char))
+tokenise reject tokenizer line col acc [] = (reverse acc, EndInput, (line, col, []))
+tokenise reject tokenizer line col acc str
+    = case scan reject [] str of
+           Just _ => (reverse acc, (EndInput, line, col, str))
+           Nothing => case getFirstMatch tokenizer str of
+                           Right (toks, line', col', rest) =>
+                               -- assert total because getFirstMatch must consume something
+                               assert_total (tokenise reject tokenizer line' col' (toks ++ acc) rest)
+                           Left reason => (reverse acc, reason, (line, col, str))
+  where
+    countNLs : List Char -> Nat
+    countNLs str = List.length (filter (== '\n') str)
+
+    getCols : List Char -> Int -> Int
+    getCols x c
+        = case span (/= '\n') x of
+               (incol, []) => c + cast (length incol)
+               (incol, _) => cast (length incol)
+
+    getFirstMatch : Tokenizer a -> List Char ->
+                    Either StopReason (List (WithBounds a), Int, Int, List Char)
+    getFirstMatch (Match lex fn) str
+        = let Just (tok, rest) = scan lex [] str
+                | _ => Left NoRuleApply
+              line' = line + cast (countNLs tok)
+              col' = getCols tok col
+              tok' = MkBounded (fn (fastPack (reverse tok))) False line col line' col' in
+              Right ([tok'], line', col', rest)
+    getFirstMatch (Compose begin mapBegin tagger middleFn endFn mapEnd) str
+        = do let Just (beginTok, rest) = scan begin [] str
+                 | _ => Left NoRuleApply
+             let line' = line + cast (countNLs beginTok)
+             let col' = getCols beginTok col
+             let beginTok' = fastPack $ reverse beginTok
+             let tag = tagger beginTok'
+             let beginTok'' = MkBounded (mapBegin beginTok') False line col line' col'
+             let middle = middleFn tag
+             let end = endFn tag
+             let (midToks, (reason, line'', col'', rest'')) =
+                    tokenise end middle line' col' [] rest
+             case reason of
+                  reason@(ComposeNotClosing _ _) => Left reason
+                  _ => Right ()
```

Comment:
What was the goal of this line? If i'm not mistaken, this desugars to
```
_<- case reason of
         reason@(ComposeNotClosing _ _) => Left reason
         _ => Right ()
```

which doesn't do anything

Reply:
I want to shortcut and return ComposeNotClosing while to do nothing on the other variances.

Reply:
I think you're looking for 

```
case reason of 
  (ComposeNotClosing _ _) ⇒ Left reason
  _ ⇒ ?restOfProgram
  ```

Reply:
I rather the way below, but it's not implemented internally:
```
let (midToks, (EndInput | NoRuleApply, line'', col'', rest'')) =
                    tokenise end middle line' col' [] rest
    | reason => Left reason
```

Reply:
Ah yes that would be nice, maybe it could be recorded in an issue and we could open a PR to implement it

Reply:
This is linked to this https://github.com/idris-lang/Idris2/pull/1034#discussion_r575457543

whether you keep `let` or `do` doesn't matter in terms of functionality. But here we want to short circuit the control flow. 


This is what we have:
```

do let …
   _ <- case reason of -- this line does not do anything
             reason@(ComposeNotClosing _ _) => Left reason
             _ => Right ()
   let …
```

This is what we would like

```
let Just (beginTok, rest) = scan begin [] str
      | _ => Left NoRuleApply
    …
    (midToks, (EndInput | NoRuleApply, line'', col'', rest'')) =
                    tokenise end middle line' col' [] rest
      | (_, (reason, _, _, _) => Left reason
    …
```

but it doesn't exist so here we what we need to do, either:

```
let Just (beginTok, rest) = scan begin [] str
      | _ => Left NoRuleApply
    … in
    case reason of
         Left (ComposeNotClosing _ _ _ ) = > Left reason
         Right _ = > let …
```

Or

```
do let (Just (beginTok, rest)) = scan begin [] str
     | _ => Left NoRuleApply
   let …
   case reason of
        Left (ComposeNotClosing _ _ _ ) = > Left reason
        Right _ = > do let …
```

Reply:
Thank you, that's clear enough. `ComposeNotClosing` is akin to `commit`, which stops trying other branches.

Reply:
I think it became worse....
```idris
    getFirstMatch (Compose begin mapBegin tagger middleFn endFn mapEnd) str
        = let Just (beginTok, rest) = scan begin [] str
                 | _ => Left NoRuleApply
              line' = line + cast (countNLs beginTok)
              col' = getCols beginTok col
              beginTok' = fastPack $ reverse beginTok
              tag = tagger beginTok'
              beginTok'' = MkBounded (mapBegin beginTok') False line col line' col'
              middle = middleFn tag
              end = endFn tag
              (midToks, (reason, line'', col'', rest'')) =
                    tokenise end middle line' col' [] rest in
              case reason of
                  reason@(ComposeNotClosing _ _) => Left reason
                  _ => let Just (endTok, rest''') = scan end [] rest''
                         | _ => Left $ ComposeNotClosing (line, col) (line', col')
                       line''' = line'' + cast (countNLs endTok)
                       col''' = getCols endTok col''
                       endTok' = fastPack $ reverse endTok
                       endTok'' = MkBounded (mapEnd endTok') False line'' col'' line''' col'''
                       Right ([endTok''] ++ reverse midToks ++ [beginTok''], line''', col''', rest''') in
```

Reply:
CALLBACK HELL!! isn't it?

Reply:
unfortunately I have to leave right now because of an unrelated obligation but here is what I came up with:

```
    getNext : (lexer : Lexer) ->  (line, col : Int) -> (input : List Char) -> Maybe (WithBounds String, List Char)
    getNext lexer line col str =
      let Just (token, rest) = scan lexer [] str
            | _ => Nothing
          line' = line + cast (countNLs token)
          col' = getCols token col
          tokenStr = fastPack $ reverse token
          parsedToken = MkBounded tokenStr False line' col' line' col'
       in pure (parsedToken, rest)

    getFirstMatch : Tokenizer a -> List Char ->
                    Either StopReason (List (WithBounds a), Int, Int, List Char)
    getFirstMatch (Match lex fn) str
        = let Just (bounded@(MkBounded tok' _ line' col' _ _), rest) = getNext lex line col str
                | _ => Left NoRuleApply
           in Right ([map fn bounded], line', col', rest)
    getFirstMatch (Compose begin mapBegin tagger middleFn endFn mapEnd) str
        = let Just (open@(MkBounded beginTok' _ line' col' _ _),  rest) = getNext begin line col str
                | _ => Left NoRuleApply
              tag = tagger beginTok'
              middle = middleFn tag
              end = endFn tag
              (midToks, (reason, line'', col'', rest'')) =
                    tokenise end middle line' col' [] rest
              in case reason of
                      (ComposeNotClosing _ _) => Left reason
                      _ => let Just (close@(MkBounded entTok'' _ line''' col''' _ _), rest''') = getNext end line'' col'' rest''
                                 | _ => Left $ ComposeNotClosing (line, col) (line', col')
                            in Right ([close] ++ reverse midToks ++ [open], line''', col''', rest''')

    getFirstMatch (Alt t1 t2) str
        = case getFirstMatch t1 str of
               Right result => Right result
               Left reason@(ComposeNotClosing _ _) => Left reason
               Left _ => getFirstMatch t2 str
```

it's not finished but you get the idea, does that seem compelling to you?